src/DotNetWorker.ApplicationInsights/FunctionsTelemetryModule.cs

```diff
@@ -20,18 +27,37 @@ public void Initialize(TelemetryConfiguration configuration)
 
             _listener = new ActivityListener
             {
-                ShouldListenTo = source => source.Name.StartsWith("Microsoft.Azure.Functions.Worker"),
+                ShouldListenTo = source => source.Name.StartsWith(TraceConstants.FunctionsActivitySource),
                 ActivityStarted = activity =>
                 {
                     var dependency = _telemetryClient.StartOperation<DependencyTelemetry>(activity);
-                    dependency.Telemetry.Type = "Azure.Functions";
-                    activity.SetCustomProperty("_depTel", dependency);
+                    dependency.Telemetry.Type = DependencyTypeInProc; // Required for proper rendering in App Insights.
+                    activity.SetCustomProperty(DependencyTelemetryKey, dependency);
                 },
                 ActivityStopped = activity =>
                 {
-                    var dependency = activity.GetCustomProperty("_depTel") as IOperationHolder<DependencyTelemetry>;
-                    _telemetryClient.StopOperation(dependency);
-                    dependency?.Dispose();
+                    // Check for Exceptions events
+                    foreach (ActivityEvent activityEvent in activity.Events)
+                    {
+                        TrackExceptionTelemetryFromActivityEvent(activityEvent, _telemetryClient);
```

Comment:
I am not sure we want to track exceptions separately like this. The reason you see both exception & request telemetry when using app insights is because they are tracked separately. It just happens that something is either calling `ILogger` with the exception and also the `Activity` is being tracked separately.

I also think AppInsights SDK has some code to hook into unhandled exceptions and export them.

In short: I think we should trust any exception here would be logged separately. We can also ensure we log unhandled exceptions from a function invocation - to make sure that scenario is covered.

Reply:
I tested without this -- and it had no `ExceptionTelemetry` associated. You'd get the `DependencyTelemetry` marked as "Error" but no context for what the error was.

The behavior here is very similar to what they're doing in the OTel exporter by pulling out the exception and logging it separately: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs#L37

Reply:
Yes, the exception telemetry would need to be logged separately. But traces and logs are separate telemetry items - we should keep their implementations separate. Is it feasible for us to add an exception log via `ILogger` somewhere for a failed function invocation? A separate middleware?

Reply:
I'm not sure I understand the comment. I'm doing the same thing that the App Insights OTel exporter is doing: if there's an Exception in the Activity, create an `ExceptionTelemetry`. The App Insights SDK doesn't do this automatically so we need to do it here.

If we add an `ILogger.LogError()` upon an Exception, that will log a `TraceTelemetry` in App Insights. It will also log an `ExceptionTelemetry`, but this can be disabled via options:

```
.AddApplicationInsightsLogger((ApplicationInsightsLoggerOptions options) => options.TrackExceptionsAsExceptionTelemetry = false)
```

I'm trying to keep the experience close to what OTel will have so that switching will be easier...

Reply:
Did not realize that implementation had this behavior. I don't exactly like the idea of mixing _log_ telemetry into _trace_ telemetry (with `ExceptionTelemetry` being a form of _log_). Can we include an option to enable/disable emitting `ExceptionTelemetry` from an `Activity`?

Reply:
I talk about separation of _trace_ vs _log_ from a general telemetry concept point of view (3 pillars of telemetry: traces, logs, metrics), not app insights specifically. App Insights predates this general concept (I think), so yeah you will see a bunch of conflicting naming/decisions there.

Ultimately, I am fine with whatever you want to do here. So long as we also add the explicit `.LogError(exception, ...)` in the `catch` statement I commented on separately. I do still think it is odd for App Insights exporter to be mixing logs and traces - it doesn't feel right. Right now, I am going to see it produce redundant exception telemetry. When we add the LogError that is, but that has to happen as errors should be logged regardless of if tracing is enabled or not.

Reply:
The LogError is something I'll do in a follow-up -- it's not really related to ActivitySource and it's not related to App Insights so I didn't want to bundle that together.